### PR TITLE
server: check fail eagerly

### LINKED ISSUE #175: Is there a way to detect a closed connection between sending stuff to the sink
If I try to send stuff in the sink and the connection is closed I'll get an error but I'm wondering if there's a way to detect the connection even if I don't send stuff into the sink for a long time.

I want to use this crate for a long running pub sub. In some case, a lot of time (hours) can occurs between each messages.

In this exemple I send something every 60 secs. If the client closes the connection, I only have an error the next time the 60 secs timer fires and I would like to know right away if possible.

```rust
struct TestService;

impl TestCase for TestService {
    fn sub(&self, ctx: RpcContext, _req: Empty, sink: ServerStreamingSink<Notification>) {
        println!("new connection");

        let (tx, rx) = futures::sync::mpsc::unbounded();

        thread::spawn(move || loop {
            println!("send");
            tx.unbounded_send(1).unwrap();
            thread::sleep(time::Duration::from_secs(60));
        });

        let f = rx.map(|v| {
            let mut n = Notification::new();
            n.set_value(v.to_string());
            (n, WriteFlags::default())
        }).map_err(|()| err_msg("error rx"))
            .forward(sink)
            .map(|_| ())
            .map_err(|e| {
                println!("error: {}", e);
                ()
            });

        ctx.spawn(f);
    }
}
```

### PR DESCRIPTION
**Pull request by BusyJay**
It's possible that server stream is flushing messages when the connection is closed. So we need to check if the call is still valid inside `poll_complete`.

Close #175.

### CODE REVIEW

**Comment by kennytm** [comment]
CI is failing this error looks legit.

```
---- cases::cancel::test_client_cancel_on_dropping stdout ----
thread 'cases::cancel::test_client_cancel_on_dropping' panicked at 'expected cancel, but got: RemoteStopped', tests/cases/cancel.rs:121:24

   7: tests::cases::cancel::check_cancel
             at tests/cases/cancel.rs:121
   8: tests::cases::cancel::test_client_cancel_on_dropping
             at tests/cases/cancel.rs:160
   9: tests::cases::cancel::test_client_cancel_on_dropping::{{closure}}
             at tests/cases/cancel.rs:142
```

**Comment by ice1000** [comment]
The error is due to #242

File: tests/cases/cancel.rs [hunk 1]
```diff
@@ -27,13 +31,15 @@ type RouteChatHandler =
 
 #[derive(Clone)]
 struct CancelService {
+    list_feature_listener: Arc<Mutex<Option<std_mpsc::Sender<()>>>>,
```
Comment (overvenus) [inline tests/cases/cancel.rs]: What about `Handler<Sender<()>>`?

File: tests/cases/cancel.rs [hunk 2]
```diff
@@ -222,3 +254,21 @@ fn test_server_cancel_on_dropping() {
...
     check_cancel(rx);
 }
+
+#[test]
+fn test_early_exit() {
+    let (service, client, _server) = prepare_suite();
+    let (tx, rx) = std_mpsc::channel();
+    *service.list_feature_listener.lock().unwrap() = Some(tx);
+    let rect = Rectangle::new();
+    {
```
Comment (overvenus) [inline tests/cases/cancel.rs]: Could you remove `{`?

File: tests/cases/cancel.rs (hunk 1 above)
Comment (BusyJay) [reply tests/cases/cancel.rs]: It's just a `Sender` rather than a `Box`.

**Comment by BusyJay** [comment]
PTAL

**Comment by Hoverbear** [comment]
@ice1000 did this fix #242?

**Comment by ice1000** [comment]
@Hoverbear Still happens to me

```
failures:

---- cases::cancel::test_client_cancel_on_dropping stdout ----
thread 'cases::cancel::test_client_cancel_on_dropping' panicked at 'expected cancel, but got: RemoteStopped', tests/cases/cancel.rs:121:24
note: Run with `RUST_BACKTRACE=1` for a backtrace.


failures:
    cases::cancel::test_client_cancel_on_dropping

test result: FAILED. 7 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out

```

(With the branch in #238 merged into master)

**Pull request merged**